src/ray/raylet/worker_pool.cc

```diff
@@ -1030,36 +1020,81 @@ void WorkerPool::PopWorker(const TaskSpecification &task_spec,
+                                     const std::string &serialized_runtime_env,
+                                     const std::string &serialized_runtime_env_context,
+                                     const PopWorkerCallback &dedicated_callback) {
+    PopWorkerStatus status = PopWorkerStatus::OK;
+    const auto &runtime_env_hash = task_spec.GetRuntimeEnvHash();
+    absl::optional<TaskWaitingForWorkerInfo> dedicated_task_info;
+    if (dedicated_callback) {
+      dedicated_task_info = TaskWaitingForWorkerInfo{task_spec.TaskId(), std::move(dedicated_callback)};
+    }
+
+    Process proc = StartWorkerProcess(
+        task_spec.GetLanguage(), rpc::WorkerType::WORKER, task_spec.JobId(),
+        dedicated_task_info,
+        &status,
+        dynamic_options, runtime_env_hash, serialized_runtime_env,
+        task_spec.OverrideEnvironmentVariables(), serialized_runtime_env_context,
+        allocated_instances_serialized_json);
+    if (status == PopWorkerStatus::OK) {
+      RAY_CHECK(proc.IsValid());
+      WarnAboutSize();
+    } else {
+      // TODO(guyang.sgy): Wait until a worker is pushed or a worker can be started If
+      // startup concurrency maxed out or job not started.
+      TriggerAsyncCallbacksForFailedWorkerStart(state, runtime_env_hash, dedicated_callback, status);
+    }
+  };
+  std::vector<std::string> dynamic_options = {};
+  PopWorkerCallback dedicated_callback = nullptr;
+  if (requires_dedicated_worker) {
+    dynamic_options = task_spec.DynamicWorkerOptions();
+    dedicated_callback = callback;
+  }
+
+  // Create runtime env.
+  if (task_spec.HasRuntimeEnv()) {
+    agent_manager_->CreateRuntimeEnv(
+        task_spec.JobId(), task_spec.SerializedRuntimeEnv(),
+        [this, start_worker_process_fn, callback, &state, task_spec, dynamic_options,
+         allocated_instances_serialized_json, requires_dedicated_worker,
+         dedicated_callback](
+            bool success, const std::string &serialized_runtime_env_context) {
+          if (success) {
+            if (!requires_dedicated_worker) {
+              state.waiting_tasks_by_env_hash[task_spec.GetRuntimeEnvHash()].emplace_back(
+                  TaskWaitingForWorkerInfo{task_spec.TaskId(), callback});
+            }
+            start_worker_process_fn(task_spec, state, dynamic_options,
+                                    task_spec.SerializedRuntimeEnv(),
+                                    serialized_runtime_env_context, dedicated_callback);
+          } else {
+            RAY_LOG(WARNING) << "Couldn't create a runtime environment for task "
+                             << task_spec.TaskId() << ". The runtime environment was "
+                             << task_spec.SerializedRuntimeEnv() << ".";
+            PopWorkerCallbackAsync(callback, nullptr, PopWorkerStatus::RuntimeEnvCreationFailed);
+          }
+        });
+  } else {
+    if (!requires_dedicated_worker) {
```

Comment:
Does it make sense to unify dedicated/vs not paths (e.g., by giving dedicated workers some unique hash id?)

Reply:
Ah good idea, I didn't like how this part turned out.